### PR TITLE
Designated initializer support

### PR DESCRIPTION
**Pull request by zygoloid**
Wording for proposal in #25.

### CODE REVIEW

File: abi-mangling.html [hunk 1]
```diff
@@ -46,9 +48,13 @@
 <tr><td>type</td> <td>d</td> <td></td> <td> builtin type double </td> </tr>
 <tr><td>oper</td> <td>d</td> <td> a </td> <td> Operator delete[] </td> </tr>
 <tr><td>oper</td> <td>d</td> <td> e </td> <td> Operator * (unary) </td> </tr>
+<tr><td>syn </td> <td>d</td> <td> i </td> <td> Designated field initializer </td> </tr>
 <tr><td>oper</td> <td>d</td> <td> l </td> <td> Operator delete </td> </tr>
+<tr><td>oper</td> <td>d</td> <td> t </td> <td> Member access (.) </td> </tr>
 <tr><td>oper</td> <td>d</td> <td> v </td> <td> Operator / </td> </tr>
 <tr><td>oper</td> <td>d</td> <td> V </td> <td> Operator /= </td> </tr>
+<tr><td>syn </td> <td>d</td> <td> x </td> <td> Designated array initializer </td> </tr>
+<tr><td>syn </td> <td>d</td> <td> X </td> <td> Designated array range initializer </td> </tr>
```
Comment (rjmccall) [inline abi-mangling.html]: I did not realize this table was here.  It is really, really badly out of date.

File: abi.html [hunk 2]
```diff
@@ -4962,8 +4962,8 @@
...
                ::= cv &lt;<a href="#mangle.type">type</a>&gt; &lt;<a href="#mangle.expression">expression</a>&gt;                               # type (expression), conversion with one argument
                ::= cv &lt;<a href="#mangle.type">type</a>&gt; _ &lt;<a href="#mangle.expression">expression</a>&gt;* E                          # type (expr-list), conversion with other than one argument
-               ::= tl &lt;<a href="#mangle.type">type</a>&gt; &lt;<a href="#mangle.expression">expression</a>&gt;* E                            # type {expr-list}, conversion with braced-init-list argument
-               ::= il &lt;<a href="#mangle.expression">expression</a>&gt;* E                                    # {expr-list}, braced-init-list in any other context
+               ::= tl &lt;<a href="#mangle.type">type</a>&gt; &lt;<a href="#mangle.expression">braced-expression</a>&gt;* E                     # type {expr-list}, conversion with braced-init-list argument
+               ::= il &lt;<a href="#mangle.expression">braced-expression</a>&gt;* E                            # {expr-list}, braced-init-list in any other context
```
Comment (rjmccall) [inline abi.html]: Please update these links to "#mangle.braced-expression".

File: abi.html [hunk 3]
```diff
@@ -5029,6 +5029,12 @@
 		 ::= L &lt;<a href="#mangle.type">type</a>&gt; &lt;<i>real-part</i> <a href="#mangle.float">float</a>&gt; _ &lt;<i>imag-part</i> <a href="#mangle.float">float</a>&gt; E   # complex floating point literal (C 2000)
                  ::= L &lt;<a href="#mangle.mangled-name">mangled-name</a>&gt; E                                 # external name
 
+  &lt;<a name="mangle.braced-expression">braced-expression</a>&gt; ::= &lt;<a href="#mangle.expression">expression</a>&gt;
+                      ::= di &lt;<i>field</i> <a href="#mangle.source-name">source-name</a>&gt; &lt;<a href="#mangle.braced-expression">braced-expression</a>&gt;    # .name = expr
+                      ::= dx &lt;<i>index</i> <a href="#mangle.expression">expression</a>&gt; &lt;<a href="#mangle.braced-expression">braced-expression</a>&gt;     # [expr] = expr
+                      ::= dX &lt;<i>range begin</i> <a href="#mangle.expression">expression</a>&gt; &lt;<i>range end</i> <a href="#mangle.expression">expression</a>&gt; &lt;<a href="#mangle.braced-expression">braced-expression</a>&gt;
+                                                                        # [expr ... expr] = expr
+
```
Comment (rjmccall) [inline abi.html]: Should we add a paragraph spelling out how this applies to nested designators, or do we think that's obvious enough from the grammar?

File: abi.html (hunk 2 above)
Comment (zygoloid) [inline abi.html]: Done.

File: abi.html (hunk 3 above)
Comment (zygoloid) [inline abi.html]: I think the fact the braced-expression production recurses back to itself is somewhat self-explanatory, but I added some extra text to remove any ambiguity regarding the order in which the designators in a designator-list should be mangled.

Comment (rjmccall) [inline abi.html]: Looks good, thanks.

**Comment by rjmccall** [comment]
Latest changes are editorial, and there's been plenty of time for feedback.  Merging.

**Pull request merged**